testplan/runnable/interactive/reloader.py

```diff
@@ -1,219 +1,530 @@
+        self._init_time = time.time()
 
-def _get_module_suites(module, suite_dict):
-    """
-    Get suite objects to update by class name (unique in one module).
-
-    :param module: Module to search for test suites.
-    :type module: ``types.ModuleType``
-    :param suite_dict: Dict of test suites.
-    :type suite_dict: ``Dict``
-    :return: Dict of suites in the current module.
-    :rtype: ``Dict``
-    """
-    module_suites = {}
-    for attr_name in dir(module):
-        attr = getattr(module, attr_name)
-        if _is_testsuite(attr):
-            module_suites[attr.__name__] = suite_dict.get(attr, [])
-    return module_suites
+    def reload(self, tests, rebuild_dependencies=False):
+        """
+        Reload code and update testsuites in given tests with new code.
 
+        :param tests: Iterable of Tests (e.g. MultiTest).
+        :type tests: ``Iterable[Test]``
+        :param rebuild_dependencies: Hard re-calculate all file dependencies.
+        :type rebuild_dependencies: ``bool``
+        """
+        start_time = time.time()
+        suite_instances = self._suites_by_class(tests)
+
+        if rebuild_dependencies:
+            self._reload_dirs, self._dep_graph, self._watched_modules = (
+                self._build_dependencies(self._extra_deps))
+
+        modified_modules = self._modified_modules
+        if modified_modules:
+            self.logger.debug(
+                "Watched files have been modified - reloading dependencies.")
+            self._reload_modified_modules(modified_modules, suite_instances)
+            self.logger.info("Took %.2f seconds to reload dependencies.",
+                             time.time() - start_time)
+        else:
+            self.logger.debug("No watched files have been modified.")
 
-def _update_suites(module, module_suites):
-    """
-    Update suite objects to new classes.
+    def _build_dependencies(self, extra_deps):
+        """
+        Build a list of directories to reload code from and a tree of
+        dependencies.
 
-    :param module: Module to update suite object for.
-    :type module: ``types.ModuleType``
-    :param module_suites: Suites in the corresponding module.
-    :type module_suites: ``Dict``
-    """
-    for attr_name in dir(module):
-        attr = getattr(module, attr_name)
-        if _is_testsuite(attr):
-            for suite_obj in module_suites.get(attr.__name__, []):
-                suite_obj.__class__ = attr
-                suite.set_testsuite_testcases(suite_obj)
+        :param extra_deps: Modules to register as extra dependencies to reload,
+            despite not being directly imported by __main__.
+        :type extra_deps: ``Iterable[ModuleType]``
+        """
+        main_module_file = sys.modules['__main__'].__file__
+        if not main_module_file:
+            raise RuntimeError(
+                "Can only use interactive reloader when the __main__ module "
```

Comment:
I saw originally _get_reload_dirs() allows absence of __main__ module, then we just reload from pwd() - why are we adding this constraint?

Reply:
There needs to be a main module to be the root of the dependency graph. It would significantly complicate the logic otherwise, since the dependency graph would not have a single root. The only use-case where there wouldn't be a __main__ module is if a user is importing testplan from the command-line interpreter and defining a testplan there dynamically (instead of just running your normal testplan script with -i) - I'm not sure if this is really something people would want to do.

We can always support no __main__ as a future enhancement, but for now I think this is simplest.

The old OSS reloader didn't have this restriction because it doesn't build a graph of dependencies, it just finds all modules in sys.modules that are locally and reloads them in arbitrary order when modified without considering which module depends on which other.